src/main/scala/firrtl/transforms/Dedup.scala

```diff
@@ -66,35 +92,124 @@ class DedupModules extends Transform with PreservesAll[Transform] {
+      {
+        c.modules.size match {
+          case 0 => (n, 0.0)
+          case 1 => (n, 1.0)
+          case d => (n, i.toDouble / (d - 1))
+        }
+      }
+    }.toMap
+    val dedupAnnotations = c.modules.map(_.name).map(ct.module).flatMap { case mt@ModuleTarget(c, m) =>
+      dedupMap.get(m) match {
+        case None => Nil
+        case Some(module: DefModule) =>
+          val paths = instanceGraph.findInstancesInHierarchy(m)
+          // If dedupedAnnos is exactly annos, contains is because dedupedAnnos is type Option
+          val newTargets = paths.map { path =>
+            path.foldLeft(ct.module(c): IsModule) { case (relPath, WDefInstance(_, name, mod, _)) =>
+              if(mod == c) CircuitTarget(c).module(c) else relPath.instOf(name, mod)
+            }
+          }
+          // Add all relative paths to referredModule to map to new instances
+          def addRecord(original: IsMember, m: IsMember): Unit = m match {
+            case x: ModuleTarget =>
+              instanceify.record(x, original)
+            case x: IsComponent =>
+              instanceify.record(x, original)
+              addRecord(original, x.stripHierarchy(1))
+          }
+          // Instanceify deduped Modules!
+          if(dedupCliques(module.name).size > 1) {
+            newTargets.foreach(t => addRecord(t, t))
+          }
+          // Return Deduped Results
+          if(newTargets.size == 1) {
+            Seq(DedupedResult(mt, newTargets.headOption, moduleName2Index(m)))
+          } else Nil
       }
-    )
+    }
+
 
-    (InferTypes.run(c.copy(modules = dedupedModules)), renameMap)
+    (InferTypes.run(c.copy(modules = dedupedModules)), instanceify.andThen(renameMap).andThen(deletedASTModules), dedupAnnotations.toList)
   }
 }
 
 /** Utility functions for [[DedupModules]] */
 object DedupModules {
+  def fastSerializedHash(s: Statement): Int ={
+    def serialize(builder: StringBuilder, nindent: Int)(s: Statement): Unit = s match {
+      case Block(stmts) => stmts.map {
+        val x = serialize(builder, nindent)(_)
+        builder ++= "\n"
+        x
+      }
+      case Conditionally(info, pred, conseq, alt) =>
+        builder ++= ("  " * nindent)
+        builder ++= s"when ${pred.serialize} :"
+        builder ++= info.serialize
+        serialize(builder, nindent + 1)(conseq)
+        builder ++= "\n" + ("  " * nindent)
+        builder ++= "else :\n"
+        serialize(builder, nindent + 1)(alt)
+      case Print(info, string, args, clk, en) =>
+        builder ++= ("  " * nindent)
+        val strs = Seq(clk.serialize, en.serialize, string.string) ++
+          (args map (_.serialize))
+        builder ++= "printf(" + (strs mkString ", ") + ")" + info.serialize
+      case other: Statement =>
+        builder ++= ("  " * nindent)
+        builder ++= other.serialize
+    }
+    val builder = new mutable.StringBuilder()
+    serialize(builder, 0)(s)
+    builder.hashCode()
+  }
```

Comment:
Not as part of this PR, but we need to do something like this for normal serialization